src/commands/filter_parser.cc

```diff
@@ -440,12 +441,36 @@ std::unique_ptr<query::Predicate> WrapPredicate(
     std::unique_ptr<query::Predicate> prev_predicate,
     std::unique_ptr<query::Predicate> predicate, bool& negate,
     query::LogicalOperator logical_operator) {
+  auto new_predicate = MayNegatePredicate(std::move(predicate), negate);
+
   if (!prev_predicate) {
-    return MayNegatePredicate(std::move(predicate), negate);
+    return new_predicate;
+  }
+
+  // Check if we can extend existing ComposedPredicate of the same type
+  query::PredicateType target_type =
+      (logical_operator == query::LogicalOperator::kAnd)
+          ? query::PredicateType::kComposedAnd
+          : query::PredicateType::kComposedOr;
+
+  if (prev_predicate->GetType() == target_type) {
+    // Flatten: Add to existing n-ary node instead of wrapping
+    auto* composed =
+        dynamic_cast<query::ComposedPredicate*>(prev_predicate.get());
+    if (composed) {
+      composed->AddChild(std::move(new_predicate));
+      return prev_predicate;  // Return extended existing node
+    }
   }
```

Comment:
We enter this IF block only on the 3rd member to be added to the Composed Structure. Let us clarify this in a comment